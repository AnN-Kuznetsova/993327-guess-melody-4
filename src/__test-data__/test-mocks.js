const MAX_ERRORS_COUNT = 3;

const genreQuestion = {
  type: `genre`,
  genre: `rock`,
  answers: [{
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `blues`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `jazz`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }],
};

const artistQuestion = {
  type: `artist`,
  song: {
    artist: `Third`,
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
  },
  answers: [{
    picture: `pic-one`,
    artist: `First`,
  }, {
    picture: `pic-two`,
    artist: `Second`,
  }, {
    picture: `pic-three`,
    artist: `Third`,
  }],
};


export {
  MAX_ERRORS_COUNT,
  artistQuestion,
  genreQuestion,
};
