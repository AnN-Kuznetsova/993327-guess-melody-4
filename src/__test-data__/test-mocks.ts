import {GameType, GenreQuestionType, ArtistQuestionType} from "../types";

const MAX_ERRORS_COUNT = 3;

const genreQuestion: GenreQuestionType = {
  type: GameType.GENRE,
  genre: `rock`,
  answers: [{
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `blues`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `jazz`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }],
};

const artistQuestion: ArtistQuestionType = {
  type: GameType.ARTIST,
  song: {
    artist: `Third`,
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
  },
  answers: [{
    picture: `pic-one`,
    artist: `First`,
  }, {
    picture: `pic-two`,
    artist: `Second`,
  }, {
    picture: `pic-three`,
    artist: `Third`,
  }],
};


export {
  MAX_ERRORS_COUNT,
  artistQuestion,
  genreQuestion,
};
