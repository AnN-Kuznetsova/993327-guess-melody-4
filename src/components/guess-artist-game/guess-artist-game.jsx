import PropTypes from "prop-types";
import React from "react";

import {ArtistQuestionsPropType} from "../../types.js";
import {withActivePlayer} from "../../hocs/with-active-player/with-active-player.jsx";


const GuessArtistGame = (props) => {
  const {question, onAnswer, renderPlayer} = props;
  const {answers, song} = question;

  const _handleUserAnswerChange = (answer) => {
    onAnswer(question, answer);
  };

  return (
    <section className="game__screen">
      <h2 className="game__title">Кто исполняет эту песню?</h2>
      <div className="game__track">
        <div className="track">
          {renderPlayer(song.src, 0)}
        </div>
      </div>

      <form className="game__artist">
        {
          answers.map((answer, index) => (
            <div className="artist" key={answer.artist + index}>
              <input
                className="artist__input visually-hidden"
                type="radio"
                name="answer"
                value={`answer-${index}`}
                id={`answer-${index}`}
                onChange={_handleUserAnswerChange.bind(null, answer)}
              />
              <label className="artist__name" htmlFor={`answer-${index}`}>
                <img className="artist__picture" src={answer.picture} alt={answer.artist} />
                {answer.artist}
              </label>
            </div>
          ))
        }
      </form>
    </section>
  );
};


GuessArtistGame.propTypes = {
  question: ArtistQuestionsPropType.isRequired,
  onAnswer: PropTypes.func.isRequired,
  renderPlayer: PropTypes.func.isRequired,
};


const GuessArtistGameWithPlayer = withActivePlayer(GuessArtistGame);


export {
  GuessArtistGame,
  GuessArtistGameWithPlayer,
};
