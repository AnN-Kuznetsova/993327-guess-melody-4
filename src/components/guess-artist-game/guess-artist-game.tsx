import * as React from "react";

import {ArtistQuestionType, AnswerArtistType} from "../../types";
import {withActivePlayer} from "../../hocs/with-active-player/with-active-player";


interface Props {
  question: ArtistQuestionType;
  onAnswer: (question: ArtistQuestionType, answer: AnswerArtistType) => void;
  renderPlayer: (src: string, id: number) => React.ReactNode;
};


const GuessArtistGame: React.FunctionComponent<Props> = (props: Props) => {
  const {question, onAnswer, renderPlayer} = props;
  const {answers, song} = question;

  const _handleUserAnswerChange = (answer) => {
    onAnswer(question, answer);
  };

  return (
    <section className="game__screen">
      <h2 className="game__title">Кто исполняет эту песню?</h2>
      <div className="game__track">
        <div className="track">
          {renderPlayer(song.src, 0)}
        </div>
      </div>

      <form className="game__artist">
        {
          answers.map((answer, index) => (
            <div className="artist" key={answer.artist + index}>
              <input
                className="artist__input visually-hidden"
                type="radio"
                name="answer"
                value={`answer-${index}`}
                id={`answer-${index}`}
                onChange={_handleUserAnswerChange.bind(null, answer)}
              />
              <label className="artist__name" htmlFor={`answer-${index}`}>
                <img className="artist__picture" src={answer.picture} alt={answer.artist} />
                {answer.artist}
              </label>
            </div>
          ))
        }
      </form>
    </section>
  );
};


const GuessArtistGameWithPlayer = withActivePlayer(GuessArtistGame);


export {
  GuessArtistGame,
  GuessArtistGameWithPlayer,
};
