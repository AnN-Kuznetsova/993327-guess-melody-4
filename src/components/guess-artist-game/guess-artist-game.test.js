import React from "react";
import renderer from "react-test-renderer";
import {GuessArtistGame} from "./guess-artist-game.jsx";
import {artistQuestion} from "../../__test-data__/test-mocks.js";
import {shallow} from "enzyme";


const props = {
  question: artistQuestion,
  onAnswer: () => {},
  renderPlayer: () => {},
};

const {answers} = props.question;

const nodeMock = {
  createNodeMock: () => {
    return {};
  }
};

const artistQuestionScreenElement = shallow(<GuessArtistGame {...props} />);


describe(`Render GuessArtistGame`, () => {
  it(`Should match with snapshot`, () => {
    const guessArtistGameSnapshot = renderer.create(
        <GuessArtistGame {...props} />, nodeMock
    ).toJSON();

    expect(guessArtistGameSnapshot).toMatchSnapshot();
  });


  it(`Should render correct answers`, () => {
    const answersArtists = answers.map((answer) => answer.artist);
    const artistElements = artistQuestionScreenElement.find(`label.artist__name`);
    const artistsAvatars = artistElements.map((artistElement) => artistElement.find(`img`));

    expect(artistElements.map((artist) => artist.text()))
      .toEqual(answersArtists);
    expect(artistsAvatars.map((artistAvatar) => artistAvatar.prop(`src`)))
      .toEqual(answers.map((answer) => answer.picture));
    expect(artistsAvatars.map((artistAvatar) => artistAvatar.prop(`alt`)))
      .toEqual(answersArtists);
  });
});
