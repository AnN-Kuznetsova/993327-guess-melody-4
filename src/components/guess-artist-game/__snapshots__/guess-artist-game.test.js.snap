// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ArtistQuestionScreen Should match with snapshot 1`] = `
<section
  className="game__screen"
>
  <h2
    className="game__title"
  >
    Кто исполняет эту песню?
  </h2>
  <div
    className="game__track"
  >
    <div
      className="track"
    >
      <button
        className="track__button track__button--pause"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
    </div>
  </div>
  <form
    className="game__artist"
  >
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="answer-0"
        name="answer"
        onChange={[Function]}
        type="radio"
        value="answer-0"
      />
      <label
        className="artist__name"
        htmlFor="answer-0"
      >
        <img
          alt="First"
          className="artist__picture"
          src="pic-one"
        />
        First
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="answer-1"
        name="answer"
        onChange={[Function]}
        type="radio"
        value="answer-1"
      />
      <label
        className="artist__name"
        htmlFor="answer-1"
      >
        <img
          alt="Second"
          className="artist__picture"
          src="pic-two"
        />
        Second
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="answer-2"
        name="answer"
        onChange={[Function]}
        type="radio"
        value="answer-2"
      />
      <label
        className="artist__name"
        htmlFor="answer-2"
      >
        <img
          alt="Third"
          className="artist__picture"
          src="pic-three"
        />
        Third
      </label>
    </div>
  </form>
</section>
`;
